src/ripple/rpc/impl/GRPCHelpers.h

```diff
@@ -0,0 +1,97 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2020 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_RPC_GRPCHELPERS_H_INCLUDED
+#define RIPPLE_RPC_GRPCHELPERS_H_INCLUDED
+
+#include "org/xrpl/rpc/v1/get_account_info.pb.h"
+#include "org/xrpl/rpc/v1/ledger_objects.pb.h"
+#include "org/xrpl/rpc/v1/meta.pb.h"
+#include "org/xrpl/rpc/v1/transaction.pb.h"
+
+#include <ripple/app/misc/TxQ.h>
+#include <ripple/ledger/TxMeta.h>
+#include <ripple/protocol/Protocol.h>
+#include <ripple/protocol/STAmount.h>
+#include <ripple/protocol/STTx.h>
+
+#include <functional>
+
+namespace ripple {
+namespace RPC {
+
+void
+populateMeta(org::xrpl::rpc::v1::Meta& proto, std::shared_ptr<TxMeta> txMeta);
+
+void
+populateQueueData(
+    org::xrpl::rpc::v1::QueueData& proto,
+    std::map<TxSeq, TxQ::AccountTxDetails const> const& txs);
+
+void
+populateTransaction(
+    org::xrpl::rpc::v1::Transaction& proto,
+    std::shared_ptr<STTx const> txnSt);
+
+void
+populateTransactionResultType(
+    org::xrpl::rpc::v1::TransactionResult& proto,
+    TER result);
+
+void
+populateAccountRoot(
+    org::xrpl::rpc::v1::AccountRoot& proto,
+    STObject const& obj);
+
+void
+populateSignerList(org::xrpl::rpc::v1::SignerList& proto, STObject const& obj);
+
+void
+populateAmount(
+    org::xrpl::rpc::v1::CurrencyAmount& proto,
+    STAmount const& amount);
+
+template <class T>
+void
+populateProtoAmount(STAmount const& amount, T& proto)
```

Comment:
I think this is a template because protobuf each amount field can have different types. I.e. "SendMax" and "DeliverMin" don't have the same type. Making them the same type simplifies this a bit.

Reply:
That's correct that they are different types. However, all of the reused field names follow a pattern; each is a unique message with the same name as the reused field, with a single field called `value`. There are a couple of reasons I did this: defining and adding new message types is easier and less error prone, since a dev does not have to do any work to keep the types consistent across different objects. Following this pattern made my own life so much easier when defining all of the messages, and leads to more robust proto definitions. I don't think having a template here is bad in any way, and am very hesitant to deviate from the pattern I described without very good reasons.

Reply:
You may be right, I really just have to think about it. The reason I flagged it is we wouldn't do something like that in C++ code. But I also understand this isn't C++ code, so different rules apply. (I was also thinking it would make overloading by member pointer nicer, but that's minor). No action required by you, I just need to think about it.

Reply:
For what its worth, I would be open to "unwrapping" message types whose `value` field is another message type, as opposed to a scalar. Meaning instead of having `SendMax` and `DeliverMin` be separate types, I would just use `CurrencyAmount` for the type of both of those fields. The value won't be default initialized. The scalars I would still want to wrap, to avoid the default initialization at the very least.

Reply:
I'm fine with whatever you decide here (see previous comment)